src/utils/user-agent-utils.ts

```diff
@@ -0,0 +1,366 @@
+        const [rgex, matcherFn] = osMatchers[i]
+        const match = rgex.exec(user_agent)
+        const result = match && matcherFn(match)
+        if (result) {
+            return result
+        }
+    }
+    return ['', '']
+}
+
+export const detectDevice = function (user_agent: string): string {
+    if (NINTENDO_REGEX.test(user_agent)) {
+        return NINTENDO
+    } else if (PLAYSTATION_REGEX.test(user_agent)) {
+        return PLAYSTATION
+    } else if (XBOX_REGEX.test(user_agent)) {
+        return XBOX
+    } else if (/ouya/i.test(user_agent)) {
+        return 'Ouya'
+    } else if (/Windows Phone/i.test(user_agent) || /WPDesktop/.test(user_agent)) {
+        return 'Windows Phone'
+    } else if (/iPad/.test(user_agent)) {
+        return IPAD
+    } else if (/iPod/.test(user_agent)) {
+        return 'iPod Touch'
+    } else if (/iPhone/.test(user_agent)) {
+        return 'iPhone'
+    } else if (/(watch)(?: ?os[,/]|\d,\d\/)[\d.]+/i.test(user_agent)) {
+        return APPLE_WATCH
+    } else if (BLACKBERRY_REGEX.test(user_agent)) {
+        return BLACKBERRY
+    } else if (/(kobo)\s(ereader|touch)/i.test(user_agent)) {
+        return 'Kobo'
+    } else if (/Nokia/i.test(user_agent)) {
+        return 'Nokia'
+    } else if (
+        // Kindle Fire without Silk / Echo Show
+        /(kf[a-z]{2}wi|aeo[c-r]{2})( bui|\))/i.test(user_agent) ||
+        // Kindle Fire HD
+        /(kf[a-z]+)( bui|\)).+silk\//i.test(user_agent)
+    ) {
+        return 'Kindle Fire'
+    } else if (/(Android|ZTE)/i.test(user_agent)) {
+        if (
+            !/Mobile/.test(user_agent) ||
+            /(9138B|TB782B|Nexus [97]|pixel c|HUAWEISHT|BTV|noble nook|smart ultra 6)/i.test(user_agent)
+        ) {
+            if (
+                (/pixel[\daxl ]{1,6}/i.test(user_agent) && !/pixel c/i.test(user_agent)) ||
+                /(huaweimed-al00|tah-|APA|SM-G92|i980|zte|U304AA)/i.test(user_agent) ||
+                (/lmy47v/i.test(user_agent) && !/QTAQZ3/i.test(user_agent))
+            ) {
+                return ANDROID
+            }
+            return ANDROID_TABLET
+        } else {
+            return ANDROID
+        }
+    } else if (/(pda|mobile)/i.test(user_agent)) {
+        return 'Generic mobile'
+    } else if (/tablet/i.test(user_agent) && !/tablet pc/i.test(user_agent)) {
+        return 'Generic tablet'
+    } else {
+        return ''
+    }
+}
+
+export const detectDeviceType = function (user_agent: string): string {
+    const device = detectDevice(user_agent)
+    if (
+        device === IPAD ||
+        device === ANDROID_TABLET ||
+        device === 'Kobo' ||
+        device === 'Kindle Fire' ||
+        device === 'Generic tablet'
+    ) {
+        return 'Tablet'
+    } else if (device === NINTENDO || device === XBOX || device === PLAYSTATION || device === 'Ouya') {
+        return 'Console'
+    } else if (device === APPLE_WATCH) {
+        return 'Wearable'
+    } else if (device) {
+        return 'Mobile'
```

Comment:
Nitpick but you have some constants available for these, might bring it down by a few more bytes?

Reply:
Ah, that's fair... I stopped replacing things so I could run a build to check it was actually a working approach. There's definitely more duplication in here